### PR TITLE
Upgrade webpack to version 5

### LINKED ISSUE #21647: Build Tooling: Configure Webpack to skip Node polyfills
Previously: #13386

**Context:**

Webpack will no longer polyfill Node modules automatically in the upcoming v5 release:

https://github.com/webpack/changelog-v5/blob/master/README.md#automatic-nodejs-polyfills-removed

**Task:**

Configure default Webpack configurations to disable `node` polyfills as `false`:

https://webpack.js.org/configuration/node/#node

**Motivation:**

Simplify a future upgrade to Webpack 5 by avoiding to put ourselves in the situation of migrating away from the automatic polyfills, by forbidding their introduction in the first place.

**Future Considerations:**

If we choose to upgrade to Webpack 5 before this task is implemented, the work would no longer be necessary.

### PR DESCRIPTION
**Pull request by jsnajdr**
Closes https://github.com/WordPress/gutenberg/issues/21647.

This is a draft PR that upgrades webpack to version 5. There are 18 small commits that make independent changes that are needed. Most can be merged as separate PRs, to make things compatible between v4 and v5 before actually doing the upgrade.

Migration guide:
https://webpack.js.org/migrate/5/

Cc @scinos who's also interested in webpack 5 but is not a member of this repo, so I cannot add him as a reviewer 🙂

### CODE REVIEW

**Comment by sgomes** [comment]
> Disable Babel runtime ES modules until their imports get fixed

They should now [be fixed](https://href.li/?https://github.com/babel/babel/pull/10853).

**Comment by gziolo** [comment]
> Cc @scinos who's also interested in webpack 5 but is not a member of this repo, so I cannot add him as a reviewer 🙂

@youknowriad, can you fix it? I'm sure Sergio has enough code contributions already and it was missed.

File: webpack.config.js
Comment (gziolo) [inline webpack.config.js:64]: I assume we will need to apply the same changes in the shared config in `@wordpress/scripts`. I can take care of it later, once the build process for Gutenberg is polished.

File: package.json
Comment (gziolo) [inline package.json:203]: Flagging this one as safe to remove:

```suggestion
		"webpack-cli": "4.1.0",
```

Unless we will start using it explicitly in the project outside of `@wordpress/scripts`.

File: webpack.config.js
Comment (jsnajdr) [reply webpack.config.js:64]: Yep, at this moment, this doesn't even have the desired effect because https://github.com/webpack/webpack/pull/11754 wasn't released yet. So it's hard to test properly.

File: package.json
Comment (jsnajdr) [reply package.json:203]: Good catch, I had to add it locally because I wanted to run webpack from command line 🙂

File: packages/dependency-extraction-webpack-plugin/tsconfig.json
Comment (sirreal) [inline packages/dependency-extraction-webpack-plugin/tsconfig.json:8]: I'm OK with removing this in the name of progress, but I'm also willing to help if we can keep the generated types.

The API of the plugins will remain the same, correct? In that case maybe we can generate and commit the types (compile and move build-types to types).

We need to delete `types` from the package.json if we remove them, or updated it if we move and commit the types.

Comment (jsnajdr) [reply packages/dependency-extraction-webpack-plugin/tsconfig.json:8]: I've been thinking about the why's of types in this plugin and my conclusion is:

Actually implementing the plugin in TypeScript, and ensuring its source is typed correctly, is super difficult. To make the plugin compatible with both webpack 4 and 5, I'd need to import two different typings into the same source, and use them at the same time:
```js
if ( webpack.version.startsWith( '4.' ) ) {
  // do stuff with webpack 4 types
} else {
  // do stuff with webpack 5 types
}
```
I don't think that's even possible.

Also, the types published by webpack are incomplete. For example, v5 doesn't publish type for `ConcatenatedModule` which we need to descend into the concatenated module's `.modules` list: it can contain `ExternalModule`s that got concatenated into one moduled and removed from the bundle.

Then we need to be to check if a `Module` is an instance of `ExternalModule` or `ConcatenatedModule`, before accessing their `.externalType` or `.modules` fields type-safely. But we can't do
```js
if ( module instanceof ConcatenatedModule )
```
because the `ConcatenatedModule` constructor is not exported by webpack. We need the physical reference to the class at runtime, not just the type declaration.

We'd need to land several webpack patches like this before we could continue.

At the same time, what we really want is probably not implementing the plugin in TypeScript, but just to publish the types declaration so that _webpack config authors_ can write their config in TypeScript.

For that, a pair of `index.js` and `index.d.ts` files, without any build step, looks ideal to me. What do you think?

Comment (sirreal) [reply packages/dependency-extraction-webpack-plugin/tsconfig.json:8]: That all seems reasonable and I'd be happy with what you propose 👍 

Thanks for sharing the detailed thoughts.

File: webpack.config.js
Comment (jsnajdr) [reply webpack.config.js:67]: A month later, `target: 'browserslist'` now works perfectly after a few more webpack 5 bugfix releases. I also added the `target` config to the shared webpack config in `packages/scripts`.

File: package.json
Comment (jsnajdr) [reply package.json]: `webpack-cli` removed from `package.json`. Now it's only in `packages/scripts/package.json`.

**Comment by gziolo** [comment]
It looks two PRs that are open:
- Babel upgrade
- webpack plugin for dependencies

Should bring this PR to a tiny size when you exclude the lock file changes 😃

**Comment by jsnajdr** [comment]
> Should bring this PR to a tiny size when you exclude the lock file changes

That's true, it will be ridiculously simple after these two PRs are merged. Change version in `package.json` from 4 to 5 and not much more than that 🙂 

Building the Gutenberg libraries is fairly straightforward, much simpler than a typical SPA with async-loaded chunks and many other optimizations.

File: packages/library-export-default-webpack-plugin/package.json
Comment (gziolo) [inline packages/library-export-default-webpack-plugin/package.json:36]: I'm surprised it works with webpack 5. It's the hackiest plugin we created to this date 😅

**Review by gziolo — COMMENTED** [review]
It looks like for `@wordpress/scripts` it's going to be a breaking change.

What's left to consider this PR ready to merge?

It would be great to bundle this PR with other breaking changes and publish in one release to avoid multiple major versions of `@wordpress/scripts` in a short period of time.

**Comment by jsnajdr** [comment]
After many of the changes merged separately, especially the Babel upgrade and `dependency-extraction-webpack-plugin` update, this main PR is fairly small and trivial now: change version numbers from 4 to 5 in a few `package.json`s, add `target: 'browserslist'` to the webpack config, update chunk hashes in Jest snapshots...

Ready to review and merge now.

**Comment by jsnajdr** [comment]
> It looks like for `@wordpress/scripts` it's going to be a breaking change.

Shall I add a notice to its changelog?

> What's left to consider this PR ready to merge?

There's still one thing that is not very nice. When building Gutenberg with the upgraded webpack, it reports several deprecation warnings, mostly in our plugins:

```
(node:97788) [DEP_WEBPACK_MAIN_TEMPLATE_ASSET_PATH] DeprecationWarning: MainTemplate.hooks.assetPath is deprecated (use Compilation.hooks.assetPath instead)
(node:97788) [DEP_WEBPACK_MAIN_TEMPLATE_RENDER_WITH_ENTRY] DeprecationWarning: MainTemplate.hooks.renderWithEntry is deprecated (use JavascriptModulesPlugin.getCompilationHooks().render instead)
(node:97788) [DEP_WEBPACK_CHUNK_TEMPLATE_RENDER_WITH_ENTRY] DeprecationWarning: ChunkTemplate.hooks.renderWithEntry is deprecated (use JavascriptModulesPlugin.getCompilationHooks().render instead)
(node:97788) [DEP_WEBPACK_MAIN_TEMPLATE_HASH] DeprecationWarning: MainTemplate.hooks.hash is deprecated (use Compilation.hooks.fullHash instead)
(node:97788) [DEP_WEBPACK_EXTERNALS_FUNCTION_PARAMETERS] DeprecationWarning: The externals-function should be defined like ({context, request}, cb) => { ... }
(node:97788) [DEP_WEBPACK_CHUNK_ENTRY_MODULE] DeprecationWarning: Chunk.entryModule: Use new ChunkGraph API
(node:97788) [DEP_WEBPACK_COMPILATION_CACHE] DeprecationWarning: Compilation.cache was removed in favor of Compilation.getCache()
(node:97788) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
	Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
	Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
(node:97788) [DEP_WEBPACK_CHUNK_MODULES_ITERABLE] DeprecationWarning: Chunk.modulesIterable: Use new ChunkGraph API
(node:97788) [DEP_WEBPACK_DEPRECATION_ARRAY_TO_SET_PUSH] DeprecationWarning: chunk.files was changed from Array to Set (using Array method 'push' is deprecated)
```

To remove these warnings, we would have to abandon compatibility with webpack 4 and use APIs that are only in v5, or to basically write two versions of each plugin, with code like:
```js
if (webpack.version.startsWith( '4.' ) ) {
  ...
} else {
  ...
}
```

I think we can stop supporting webpack 4 and release a new major versions of affected plugins, but I'd rather do it only after this upgrade is done and settled.

File: packages/library-export-default-webpack-plugin/package.json
Comment (jsnajdr) [reply packages/library-export-default-webpack-plugin/package.json:36]: It's a very simple one though 🙂 

The plugin may be not needed at all: webpack already has an `output: { libraryExport: 'default' }` option that does the same thing. And the implementation is very similar to yours -- just append `.default` at the end 🙂.

The only caveat is that we want to export the `default` only for certain libraries/entrypoints, while the option is global 🙁

**Comment by gziolo** [comment]
> > It looks like for @wordpress/scripts it's going to be a breaking change.
> 
> Shall I add a notice to its changelog?

That would be fantastic.

> There's still one thing that is not very nice. When building Gutenberg with the upgraded webpack, it reports several deprecation warnings, mostly in our plugins:

I don't mind seeing those warnings for some time as long as it's tracked in its own issue.

> I think we can stop supporting webpack 4 and release a new major versions of affected plugins, but I'd rather do it only after this upgrade is done and settled.

Totally agree. Let's ensure first that webpack 5 works without any issues first.

**Review by gziolo — APPROVED** [review]

**Comment by gziolo** [comment]
It looks like Storybook doesn't work at all when webpack 5 is installed:

https://github.com/storybookjs/storybook/issues/9216

**Comment by jsnajdr** [comment]
Thanks @gziolo for finishing and merging this upgrade :tada:

> It looks like Storybook doesn't work at all when webpack 5 is installed:

Does that mean that Storybook is currently broken in Gutenberg?

**Comment by gziolo** [comment]
> Does that mean that Storybook is currently broken in Gutenberg?

Yes, it is broken. See: https://github.com/WordPress/gutenberg/runs/1640682536

Screen Shot 2021-01-04 at 09 19 36

As a temporary solution, we need to find a creative way to run Storybook with webpack 4 ...

**Comment by gziolo** [comment]
`thread-loader` and `webpack-livereload-plugin` don't quite work with `@wordpress/scripts`. I'm trying to fix all that in another PR, related commit:
https://github.com/WordPress/gutenberg/pull/27821/commits/4269f31110a502b01b3926be1c46a8cd31a75e81

**Comment by gziolo** [comment]
`wp-scripts start` errors with the scaffolded block. Steps to reproduce:

```
npx wp-create-block --no-wp-scripts
$ cd esnext-example 
$ ../node_modules/.bin/wp-scripts start
(node:72132) [DEP_WEBPACK_WATCH_WITHOUT_CALLBACK] DeprecationWarning: A 'callback' argument need to be provided to the 'webpack(options, callback)' function when the 'watch' option is set. There is no way to handle the 'watch' option without a callback.
(Use `node --trace-deprecation ...` to show where the warning was created)
Live Reload listening on port 35729

[webpack-cli] Compilation starting...
[webpack-cli] Uncaught exception: Error: MainTemplate.hooks.startup has been removed (use RuntimeGlobals.startup instead)
[webpack-cli] Error: MainTemplate.hooks.startup has been removed (use RuntimeGlobals.startup instead)
    at Object.tap (/Users/gziolo/Projects/gutenberg/node_modules/webpack/lib/MainTemplate.js:100:12)
    at LiveReloadPlugin.applyCompilation (/Users/gziolo/Projects/gutenberg/node_modules/webpack-livereload-plugin/index.js:164:42)
    at Hook.eval [as call] (eval at create (/Users/gziolo/Projects/gutenberg/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:104:1)
    at Hook.CALL_DELEGATE [as _call] (/Users/gziolo/Projects/gutenberg/node_modules/webpack/node_modules/tapable/lib/Hook.js:14:14)
    at Compiler.newCompilation (/Users/gziolo/Projects/gutenberg/node_modules/webpack/lib/Compiler.js:943:26)
    at /Users/gziolo/Projects/gutenberg/node_modules/webpack/lib/Compiler.js:984:29
    at Hook.eval [as callAsync] (eval at create (/Users/gziolo/Projects/gutenberg/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/Users/gziolo/Projects/gutenberg/node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)
    at Compiler.compile (/Users/gziolo/Projects/gutenberg/node_modules/webpack/lib/Compiler.js:979:28)
    at /Users/gziolo/Projects/gutenberg/node_modules/webpack/lib/Watching.js:112:19
```

**Comment by gziolo** [comment]
I opened #27974 where I plan to revert these changes. It looks like we need to wait another couple of weeks until Storybook and all webpack plugins in use have good support for v5.

**Comment by gziolo** [comment]
It's now reverted. We need to open another PR to track progress. @jsnajdr, major props for the work done so far 🙇🏻

**Comment by jsnajdr** [comment]
The issue with Storybook is very similar (if not exactly the same) as the one that @scinos encountered in Calypso with `mini-css-extract-plugin` (reported in https://github.com/webpack-contrib/mini-css-extract-plugin/issues/668). But here we're dealing with `terser-webpack-plugin` instead.

`@storybook/core` depends on webpack 4 and has a copy in its private `node_modules` folder. There's webpack 5 in the root `node_modules`, so the v4 can't be hoisted there. Both the `@storybook/core` and the root package depend on `terser-webpack-plugin`. Both dependencies are compatible, so they share a common copy hoisted into root `node_modules`.

Now the webpack 4 compilation in `@storybook/core` requires `terser-webpack-plugin`, which in turn requires `webpack`, but in the root `node_modules` the `webpack` request resolves to the webpack 5 copy there.

The `webpack` -> `terser-webpack-plugin` -> `webpack` require chain is supposed to be a closed loop where both webpacks are the same module, but in this case they are two different modules.

A possible solution could be something like [Yarn's `nohoist` option](https://classic.yarnpkg.com/blog/2018/02/15/nohoist/), which could prevent `terser-webpack-plugin`, or any other dependencies of `@storybook/*`, from hoisting to the top. That creates a kind of "sandbox" environment for Storybook, not affected by the root `node_modules` folder. But that works only for Yarn, and also only for workspaces, not arbitrary packages in dependencies.

Another option is to manually patch the `package-lock.json` lockfile to install a copy of `terser-webpack-plugin` inside `@storybook/core`. But that's very fragile and will break on the very next `npm install` run.

**Comment by sirreal** [comment]
I want to surface another issue I know @jsnajdr was already aware of. Webpack 5 compiled sources may not be compatible with the Plugin Repository i18n string extraction. Apparently webpack 5 outputs module function calls of the form `(0, mod.exp)(params)` whereas with webpack 4 they had the form `Object(mod.exp)(params)` (I'm paraphrasing @jsnajdr in another conversation).

The webpack 5 form does not seem to be properly extracted (e.g. `(0,wpI18n.__)('translated string')`), which seems like a blocker for updating this.

cc: @swissspidy

**Comment by swissspidy** [comment]
@sirreal Sounds like we should add support for it in WP-CLI soon then!

Can you confirm that this is the correct output for webpack 5?

**Comment by simison** [comment]
See  https://github.com/wp-cli/i18n-command/pull/228

**Comment by sirreal** [comment]
> @sirreal Sounds like we should add support for it in WP-CLI soon then!
> 
> Can you confirm that this is the correct output for webpack 5?
Yes, I looked at the compiled plugin from the commit that landed this PR: https://github.com/WordPress/gutenberg/commit/85eda573b6f02da1d8e8e4c4437d1f3a820e5b54

It shows this type of usage. Here's a sample:

```
(0,xe.__)("No tags.")
(0,xe.__)("Post tags block: Tags are not available for this post type.")
(0,xe.__)("Post tags block: No post found for this block.")
…
```

**Comment by swissspidy** [comment]
Ah, looks like it's a Babel thing ten, not webpack specific

**Comment by jsnajdr** [comment]
@swissspidy The `(0,xe.__)('No tags.')` calls are created by webpack, when replacing usages of `import`-ed bindings with its own plumbing code.

In webpack 4, they used `Object(xe.__)('No tags.')`. The purpose of both these tricks is to unbind the `this` pointer of the `__` function from being `xe` and reset `this` to either `global` or `undefined`.

**Comment by gziolo** [comment]
Webpack 5 support for Storybook is still behind the flag but at least there is some progress:
https://storybook.js.org/blog/storybook-for-webpack-5/

> We’re excited to announce experimental Webpack 5 support in Storybook, the world’s most popular UI component workshop.

**Comment by gziolo** [comment]
#30347 should unblock webpack 5 upgrade, @jsnajdr are you up for giving it another try?

**Comment by jsnajdr** [comment]
> are you up for giving it another try?

OK 🙂 Hopefully in the next two weeks -- I'm a bit overwhelmed these days.

**Pull request merged**
